### PR TITLE
Change add_noise to match the target variance by scaling by its sqrt.

### PR DESCRIPTION
**Pull request by Rupt**
Resolve issue #281.

Add a test for this noise scaling which fails prior to this change.

### CODE REVIEW

**Comment by Rupt** [comment]
Edited to avoid sqrt after type casting.

File: optax/_src/transform.py [hunk 1]
```diff
@@ -717,12 +717,13 @@ def update_fn(updates, state, params=None):  # pylint: disable=missing-docstring
     treedef = jax.tree_structure(updates)
     count_inc = numerics.safe_int32_increment(state.count)
     variance = eta / count_inc**gamma
+    std = jnp.sqrt(variance)
```
Comment (mkunesch) [inline optax/_src/transform.py:720]: Nit: std is a standard abbreviation so it's perfectly fine as is, but I'd personally prefer `standard_deviation`, also to be consistent with the use of `variance` over `var`.

File: optax/_src/transform_test.py [hunk 2]
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
...
+    updates = jax.tree_map(jnp.zeros_like, self.per_step_updates)
+
+    for i in range(1, STEPS + 1):
+      updates_i, state = noise.update(updates, state)
+      updates_i_unit, state_unit = noise_unit.update(updates, state_unit)
+
+      scale = jnp.sqrt(eta / i**gamma)
+
+      def rescale(g):
+        return g * scale  # pylint:disable=cell-var-from-loop
```
Comment (mkunesch) [inline optax/_src/transform_test.py:229]: I think this is simple enough to be written as lambda in the line below. That way you also wouldn't need to disable the pylint warning or did pylint complain about something else then?

File: optax/_src/transform_test.py [hunk 3]
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
...
+    # Prepare to compare noise with a rescaled unit-variance substitute.
+    noise = transform.add_noise(eta, gamma, seed)
+    noise_unit = transform.add_noise(1.0, 0.0, seed)
+
+    params = self.init_params
+    state = noise.init(params)
+    state_unit = noise_unit.init(params)
+
+    # Check the noise itself by adding it to zeros.
+    updates = jax.tree_map(jnp.zeros_like, self.per_step_updates)
```
Comment (mkunesch) [inline optax/_src/transform_test.py:220]: Nit: you could make this more self-contained by replacing `self.per_step_updates` with `params` directly (because that's ultimately the shape we want to match).

File: optax/_src/transform_test.py [hunk 4]
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
     centralized_inputs, _ = centralizer.update(inputs, None)
     chex.assert_tree_all_close(centralized_inputs, outputs)
 
+  @parameterized.parameters(
+    (0.01, 0.55, 123),
+    (0.3, 0.54, 234),
+    (1.0, 0.56, 314),
+  )
+  def test_add_noise(self, eta, gamma, seed):
```
Comment (mkunesch) [inline optax/_src/transform_test.py:210]: Could you provide more information on the behaviour being tested? Either by changing the name to e.g. `test_add_noise_has_correct_standard_deviation` or by adding a docstring with something similar (just so that there is more information if the test fails). Thanks a lot!

**Review by mkunesch — CHANGES_REQUESTED** [review]
Looks great! Thank you so much for this bug fix and the associated test!
I just left a few very minor comments/nits.

File: optax/_src/transform.py (hunk 1 above)
Comment (Rupt) [reply optax/_src/transform.py:720]: Done.
(I chose std to match its use in stochastic_gradient_estimators, but totally agree with your argument)

File: optax/_src/transform_test.py (hunk 2 above)
Comment (Rupt) [reply optax/_src/transform_test.py:229]: Yes this is because pylint complains with the lambda form, and with this same warning. This style copies [L186](https://github.com/deepmind/optax/blob/a0d5ed5539a58cc0bfa1d5948c87d2aa3b693d92/optax/_src/transform_test.py#L186).

File: optax/_src/transform_test.py (hunk 3 above)
Comment (Rupt) [reply optax/_src/transform_test.py:220]: Nice! Done.

File: optax/_src/transform_test.py (hunk 4 above)
Comment (Rupt) [reply optax/_src/transform_test.py:210]: I've gone with `test_add_noise_has_correct_variance_scaling` since the test doesn't actually check the standard deviation, just the scale relative to eta=1, gamma=0.
Thanks for all the nice comments.

**Review by mkunesch — CHANGES_REQUESTED** [review]
Thanks a lot for making the changes! I forgot to check for `chex.all_variants` last time so I have one more comment. Otherwise looks good to me!

File: optax/_src/transform_test.py (hunk 2 above)
Comment (Rupt) [reply optax/_src/transform_test.py]: Neat solution, done with the lambda.

File: optax/_src/transform_test.py [hunk 5]
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
     centralized_inputs, _ = centralizer.update(inputs, None)
     chex.assert_tree_all_close(centralized_inputs, outputs)
 
+  @parameterized.parameters(
+    (0.01, 0.55, 123),
+    (0.3, 0.54, 234),
+    (1.0, 0.56, 314),
+  )
+  def test_add_noise_has_correct_variance_scaling(self, eta, gamma, seed):
```
[the remark this answers is not in the document]
Comment (Rupt) [reply optax/_src/transform_test.py]: Added the variants.

Agreed there is nothing new the other tests can catch. Kwargs there don't work because the parameters are reused elsewhere (seed for the other transform, eta and gamma for scale), so I've just hard coded them as variables.

**Review by mkunesch — COMMENTED** [review]
LGTM! Thanks a lot!

There is a test failure now in the (recently added) test for `noisy_sgd` in `alias_test.py` which uses `add_noise`. Do you have time to look into this? It's probably just a matter of changing the parameters so that with the slower decay of the noise it doesn't affect the test results but we should make sure this is indeed the reason.

**Comment by Rupt** [comment]
> It's probably just a matter of changing the parameters
I've rebased and changed eta to match the previous initial variance. It survives the test with the different decay rate.
 
> but we should make sure this is indeed the reason.
The test passes now (and also passed if just removing the noisy_sgd line.

**Review by mkunesch — APPROVED** [review]
Great, thanks a lot for doing this!
